File: apps/android/app/src/main/java/com/sanvya/app/ui/MoneyFormat.kt
package com.sanvya.app.ui

import com.sanvya.app.domain.money.Money
import com.sanvya.app.domain.money.format
import com.sanvya.app.domain.money.minorUnits
import com.sanvya.app.domain.money.money
import com.sanvya.app.domain.money.toMajor

/**
 * The one money formatter.
 *
 * Before this file, eleven screens each built their own:
 *
 * ```
 * NumberFormat.getCurrencyInstance(Locale("en", "IN")).apply {
 *     currency = Currency.getInstance("INR")
 * }
 * ```
 *
 * Every one of them was wrong the same three ways. The currency was pinned to
 * INR, so a USD account rendered with a rupee sign. The locale was pinned to
 * `en_IN`, so **every** currency got lakh/crore grouping — $1,25,000. And none
 * of them consulted the hide-amounts setting, which is the privacy-leak class
 * that has already shipped on web three times (PARITY_AUDIT trap 7).
 *
 * Two of the eleven were also called `formatMoney`, declared `internal` in
 * different files with *different* `maximumFractionDigits` — 0 in Loans, 2 in
 * Splits — so the same amount rendered differently on two screens.
 *
 * Everything now goes through the generated `domain.money.format`, which takes
 * its fraction digits from `minorUnits(currency)` and its grouping from the
 * currency's own locale. This mirrors iOS's `App/Components/MoneyFormat.swift`
 * exactly, so the two platforms cannot drift.
 */

/** Shown in place of an amount when hide-amounts is on. Matches web's `useMoneyFmt`. */
const val MONEY_MASK = "••••"

/**
 * Format a [Money], respecting the hide-amounts privacy setting.
 *
 * This is the native `useMoneyFmt()`. Use it for **every** amount that reaches
 * the screen — including inside charts, which is where web's leaks happened.
 */
/**
 * The divisor that turns minor units into major ones for a currency.
 *
 * For a CHART, where the value is already a Double (an average, a running
 * total) and wrapping it back into a `Money` would round it. Everywhere an
 * exact integer amount is in hand, `toMajor` is the right call instead.
 *
 * It exists so chart code stops writing `/ 100.0`, which is correct for the
 * rupee, the dollar and the euro and wrong for the yen -- a JPY chart drawn
 * that way plots every point at a hundredth of its real height, silently.
 */
fun majorScale(currency: String): Double = Math.pow(10.0, minorUnits(currency).toDouble())

/**
 * The unformatted major-unit value, for a field the user is about to edit.
 *
 * Deliberately NOT `formatMoney`: an input field must contain something the
 * user can type back, so no symbol, no grouping and no mask. It is still
 * currency-aware.
 *
 * **There were four copies of this, and three of them divided by 100.** The one
 * correct copy lived in InvestmentsViewModel with a comment explaining exactly
 * why; the other three put two extra decimal places into a JPY field and
 * dropped one from a KWD field. A helper duplicated per screen is a helper that
 * gets fixed on one screen.
 */
fun formatMajorPlain(minor: Long, currency: String): String {
    val major = toMajor(money(minor, currency))
    return if (major == Math.floor(major)) major.toLong().toString() else major.toString()
}

fun formatMoneyAware(m: Money, mask: String = MONEY_MASK): String =
    if (Prefs.amountsHidden.value) mask else format(m)

/** Convenience for the common `(minorUnits, currencyCode)` call shape. */
fun formatMoney(minor: Long, currency: String, mask: String = MONEY_MASK): String =
    formatMoneyAware(money(minor, currency), mask)

/**
 * Format without masking.
 *
 * For the rare place an amount must always be visible — an input field the user
 * is actively editing, where masking would make the field unusable rather than
 * private. **Not** for display.
 */
fun formatMoneyUnmasked(m: Money): String = format(m)

/**
 * The user's base currency, for amounts that roll up across accounts.
 *
 * Net worth, a cross-group split position, a multi-currency total: none of these
 * is *in* any one account's currency, so they report in the currency the user
 * chose. Reading it here rather than at eleven call sites is what makes the
 * setting mean something — it was write-only until now.
 */
fun baseCurrencyNow(): String = Prefs.baseCurrency.value

File: apps/ios/App/Components/MoneyFormat.swift
import Foundation
import Domain

/**
 The one money formatter.

 Before this file there were four, and every one of them was wrong in a
 different way:

 - `formatMoneyAware` (SettingsView.swift) masked correctly but hardcoded
   `"₹%.2f"`, so every non-INR amount rendered with a rupee sign.
 - `formatMoney` (LoansViewModel.swift) divided by 100, forced `en_IN`, and
   showed zero fraction digits.
 - `formatMoney` (ReceiptReviewView.swift) divided by 100 with two fraction
   digits — and, being a second module-scope declaration of the same name,
   was a hard compile error.
 - `formatCents` (DashboardView.swift) divided by 100 and hardcoded INR.

 Three of the four ignored the hide-amounts setting entirely, which is the
 privacy leak class that has already shipped on web three times (PARITY_AUDIT.md
 trap 7), and all four hardcoded ÷100, which is wrong for JPY (no minor unit)
 and BHD (three) and contradicts golden rule 1.

 Everything now goes through `Domain.format`, which is generated from
 `packages/core/money` and takes its fraction digits from `minorUnits(currency)`
 and its grouping from the currency's own locale.
 */

/// Mask shown in place of an amount when hide-amounts is on. Matches web's
/// `useMoneyFmt` default.
public let moneyMask = "••••"

/// Whether amounts are currently hidden, readable from any isolation context.
///
/// `Prefs` is `@MainActor` because it is an `ObservableObject` driving SwiftUI,
/// but some formatting happens in `@Sendable` closures with no actor at all —
/// the insight generators hand a plain formatter closure to domain code. Making
/// the formatter `@MainActor` would put those call sites permanently out of
/// reach of the masking rule, which is exactly how an amount leaks. Reading the
/// same UserDefaults key directly is thread-safe and needs no isolation, so
/// there is one setting with one meaning and no context where it cannot be
/// honoured.
public func amountsHiddenNow() -> Bool {
    UserDefaults.standard.bool(forKey: Prefs.hideKey)
}

/// Format a `Money`, respecting the hide-amounts privacy setting.
///
/// This is the native `useMoneyFmt()`. Use it for every amount that reaches the
/// screen — including inside charts, which is where web's leaks happened.
public func formatMoneyAware(_ money: Domain.Money, mask: String = moneyMask) -> String {
    amountsHiddenNow() ? mask : Domain.format(money)
}

/// Convenience for the common `(minorUnits, currencyCode)` call shape.
public func formatMoney(_ minor: Int64, _ currency: String, mask: String = moneyMask) -> String {
    formatMoneyAware(Domain.money(minor, currency), mask: mask)
}

/// Format without masking — for the rare place an amount must always be
/// legible, such as the confirmation text of a destructive action the user has
/// deliberately opened. Call sites should be few and obvious.
public func formatMoneyUnmasked(_ money: Domain.Money) -> String {
    Domain.format(money)
}

/// The user's base currency, readable without isolation — same reasoning as
/// `amountsHiddenNow()`.
public func baseCurrencyNow() -> String {
    UserDefaults.standard.string(forKey: Prefs.currencyKey) ?? FormOptions.defaultCurrency
}

/// A bare, ungrouped major-unit number for an EDITABLE field — "1200", or
/// "1200.5" when there is a fraction. Never masked: it is what the user is
/// typing, not something being shown to them.
///
/// Four identical private copies of this existed, all dividing by 100. The
/// divisor now comes from `minorUnits(currency)`, so a JPY amount is not
/// silently divided by a hundred.
public func formatMajorPlain(_ minor: Int64, currency: String = baseCurrencyNow()) -> String {
    let digits = Domain.minorUnits(currency)
    let major = Double(minor) / pow(10.0, Double(digits))
    return major == major.rounded() ? String(Int64(major)) : String(major)
}

/// Abbreviated amount for tight spaces — ₹1.5L / ₹2.3Cr for Indian-numbering
/// currencies, $1.2K / $3.4M otherwise. Masked like any other displayed amount.
///
/// Web renders these through `Intl.NumberFormat`'s `notation: "compact"`, whose
/// exact breakpoints are not reproducible here; the spec accepts the drift.
public func compactMoney(_ minor: Int64, _ currency: String) -> String {
    if amountsHiddenNow() { return moneyMask }
    let digits = Domain.minorUnits(currency)
    let major = Double(minor) / pow(10.0, Double(digits))
    let magnitude = abs(major)
    let indian = ["INR", "PKR", "LKR", "BDT", "NPR"].contains(currency.uppercased())

    let (value, suffix): (Double, String)
    if indian, magnitude >= 10_000_000 { (value, suffix) = (major / 10_000_000, "Cr") }
    else if indian, magnitude >= 100_000 { (value, suffix) = (major / 100_000, "L") }
    else if !indian, magnitude >= 1_000_000_000 { (value, suffix) = (major / 1_000_000_000, "B") }
    else if !indian, magnitude >= 1_000_000 { (value, suffix) = (major / 1_000_000, "M") }
    else if magnitude >= 1_000 { (value, suffix) = (major / 1_000, "K") }
    else { return Domain.format(Domain.money(minor, currency)) }

    let formatter = NumberFormatter()
    formatter.numberStyle = .currency
    formatter.currencyCode = currency
    formatter.locale = Locale(identifier: currencyLocales[currency] ?? "en_US")
    formatter.maximumFractionDigits = abs(value) >= 100 ? 0 : 1
    let body = formatter.string(from: NSNumber(value: value)) ?? "\(value)"
    return body + suffix
}
